File: 347-top-k-frequent-elements/347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        
        // 1.
        // unordered_map<int, int> mp;
        // for(int i=0;i<nums.size();i++){
        //     mp[nums[i]]++;
        // }
        // vector<int> ans;
        // while(ans.size() < k){
        //     int f = INT_MIN;
        //     int t;
        //     for(auto it : mp){            
        //         if(f < it.second){
        //             f = it.second;
        //             t = it.first;
        //         }
        //     }
        //     if(ans.size() < k)
        //         ans.push_back(t);
        //     mp.erase(t);
        // }
        // return ans;
        
        
      // by default the priority queue will sort the first value of pairs in descending order
        // 2. 
        unordered_map<int, int> mp;
        for(int i=0;i<nums.size();i++){
            mp[nums[i]]++;
        }
        vector<int> ans(k, 0);
        priority_queue<pair<int, int>> pq;
        for(auto it : mp){
            pair<int, int> p;
            p.first = it.second;
            p.second = it.first;
            pq.push(p);
        }
        int i=0;
        while(i<k){
            ans[i] = pq.top().second;
            pq.pop();
            i++;
        }
        return ans;
    }
};

File: 347-top-k-frequent-elements/347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int []ans=new int[k];
        Map<Integer,Integer>map=new HashMap<>();
        for(int i:nums){
            map.put(i,map.getOrDefault(i,0)+1);
        }
        PriorityQueue<int[]> pq=new PriorityQueue<>((a,b)->b[1]-a[1]);
        for(Map.Entry<Integer,Integer> entry:map.entrySet()){
            int[]a=new int[]{entry.getKey(),entry.getValue()};
            pq.add(a);
        }
        int i=0;
        while(i<k){
            ans[i]=pq.remove()[0];
            i++;
        }
        return ans;
    }
}